# NumPy Retail Analytics Assignment

## Business Scenario

You are working as a Data Analyst for **SmartMart Retail Stores**. The company operates five stores in different cities and sells six major product categories.

Management wants to analyze sales performance, expenses, profits, targets, discounts, inventory, and store-level performance using NumPy.

You are provided with two main NumPy arrays:

1. `sales_data` — monthly sales revenue generated by each store for different product categories.
2. `cost_data` — total cost incurred by each store for the same product categories.

---

# Part 1: Dataset Creation

In [ ]:
import numpy as np

# Store names
stores = np.array([
    "Bengaluru",
    "Hyderabad",
    "Chennai",
    "Mumbai",
    "Pune"
])

# Product categories
categories = np.array([
    "Electronics",
    "Groceries",
    "Clothing",
    "Furniture",
    "Beauty",
    "Sports"
])

# Sales revenue in rupees
# Rows represent stores
# Columns represent product categories

sales_data = np.array([
    [850000, 420000, 310000, 275000, 190000, 225000],
    [780000, 460000, 295000, 310000, 210000, 240000],
    [920000, 390000, 340000, 260000, 185000, 215000],
    [1050000, 510000, 375000, 345000, 250000, 290000],
    [690000, 405000, 280000, 225000, 175000, 205000]
])

# Cost incurred in rupees
# Rows represent stores
# Columns represent product categories

cost_data = np.array([
    [640000, 315000, 220000, 205000, 125000, 160000],
    [590000, 350000, 210000, 225000, 140000, 170000],
    [700000, 295000, 245000, 195000, 120000, 155000],
    [790000, 390000, 265000, 250000, 165000, 205000],
    [530000, 310000, 205000, 170000, 115000, 145000]
])

print("Stores:")
print(stores)

print("\nProduct Categories:")
print(categories)

print("\nSales Data:")
print(sales_data)

print("\nCost Data:")
print(cost_data)

Stores:
['Bengaluru' 'Hyderabad' 'Chennai' 'Mumbai' 'Pune']

Product Categories:
['Electronics' 'Groceries' 'Clothing' 'Furniture' 'Beauty' 'Sports']

Sales Data:
[[ 850000  420000  310000  275000  190000  225000]
 [ 780000  460000  295000  310000  210000  240000]
 [ 920000  390000  340000  260000  185000  215000]
 [1050000  510000  375000  345000  250000  290000]
 [ 690000  405000  280000  225000  175000  205000]]

Cost Data:
[[640000 315000 220000 205000 125000 160000]
 [590000 350000 210000 225000 140000 170000]
 [700000 295000 245000 195000 120000 155000]
 [790000 390000 265000 250000 165000 205000]
 [530000 310000 205000 170000 115000 145000]]


---

# Part 2: Basic Array Understanding

## Question 1: Array Dimensions

The finance manager wants to confirm how many stores and product categories are present in the data.

Write NumPy code to display:

* Number of dimensions in `sales_data`
* Shape of `sales_data`
* Total number of values
* Data type of the array
* Memory occupied by the array

In [ ]:
sales_ndim = sales_data.ndim
sales_shape = sales_data.shape
sales_size = sales_data.size
sales_dtype = sales_data.dtype
sales_nbytes = sales_data.nbytes

print("Number of dimensions:", sales_ndim)
print("Shape:", sales_shape)
print("Total number of values:", sales_size)
print("Data type:", sales_dtype)
print("Memory occupied:", sales_nbytes, "bytes")

Number of dimensions: 2
Shape: (5, 6)
Total number of values: 30
Data type: int64
Memory occupied: 240 bytes


---

## Question 2: Store and Category Count

Calculate the following:

* Total number of stores
* Total number of product categories
* Total number of store-category combinations

In [ ]:
num_stores = stores.shape[0]
num_categories = categories.shape[0]
num_combinations = sales_data.size

print("Total number of stores:", num_stores)
print("Total number of product categories:", num_categories)
print("Total number of store-category combinations:", num_combinations)

Total number of stores: 5
Total number of product categories: 6
Total number of store-category combinations: 30


---

## Question 3: Sales Data Type Conversion

The reporting application requires sales values in floating-point format.

Convert `sales_data` into:

* `float32`
* `float64`

Compare the memory used by both arrays.

In [ ]:
sales_float32 = sales_data.astype(np.float32)
sales_float64 = sales_data.astype(np.float64)

print("float32 array:\n", sales_float32)
print("\nfloat64 array:\n", sales_float64)

print("\nMemory used by int array (original):", sales_data.nbytes, "bytes")
print("Memory used by float32 array:", sales_float32.nbytes, "bytes")
print("Memory used by float64 array:", sales_float64.nbytes, "bytes")
print("float64 uses", sales_float64.nbytes - sales_float32.nbytes, "more bytes than float32")

float32 array:
 [[8.50e+05 4.20e+05 3.10e+05 2.75e+05 1.90e+05 2.25e+05]
 [7.80e+05 4.60e+05 2.95e+05 3.10e+05 2.10e+05 2.40e+05]
 [9.20e+05 3.90e+05 3.40e+05 2.60e+05 1.85e+05 2.15e+05]
 [1.05e+06 5.10e+05 3.75e+05 3.45e+05 2.50e+05 2.90e+05]
 [6.90e+05 4.05e+05 2.80e+05 2.25e+05 1.75e+05 2.05e+05]]

float64 array:
 [[ 850000.  420000.  310000.  275000.  190000.  225000.]
 [ 780000.  460000.  295000.  310000.  210000.  240000.]
 [ 920000.  390000.  340000.  260000.  185000.  215000.]
 [1050000.  510000.  375000.  345000.  250000.  290000.]
 [ 690000.  405000.  280000.  225000.  175000.  205000.]]

Memory used by int array (original): 240 bytes
Memory used by float32 array: 120 bytes
Memory used by float64 array: 240 bytes
float64 uses 120 more bytes than float32


---

# Part 3: Indexing and Data Extraction

## Question 4: Bengaluru Store Analysis

The Bengaluru regional manager wants to review sales for all product categories.

Extract and display the complete sales row for Bengaluru.

In [ ]:
bengaluru_index = np.where(stores == "Bengaluru")[0][0]
bengaluru_sales = sales_data[bengaluru_index]

print("Bengaluru sales across all categories:")
for cat, val in zip(categories, bengaluru_sales):
    print(f"{cat}: {val}")

Bengaluru sales across all categories:
Electronics: 850000
Groceries: 420000
Clothing: 310000
Furniture: 275000
Beauty: 190000
Sports: 225000


---

## Question 5: Mumbai Cost Analysis

Extract the total cost incurred by the Mumbai store for all product categories.

In [ ]:
mumbai_index = np.where(stores == "Mumbai")[0][0]
mumbai_cost = cost_data[mumbai_index]

print("Mumbai cost across all categories:")
for cat, val in zip(categories, mumbai_cost):
    print(f"{cat}: {val}")

Mumbai cost across all categories:
Electronics: 790000
Groceries: 390000
Clothing: 265000
Furniture: 250000
Beauty: 165000
Sports: 205000


---

## Question 6: Electronics Sales Across Stores

The electronics department wants to compare its sales performance across all five stores.

Extract the Electronics sales column from `sales_data`.

In [ ]:
electronics_index = np.where(categories == "Electronics")[0][0]
electronics_sales = sales_data[:, electronics_index]

print("Electronics sales across all stores:")
for st, val in zip(stores, electronics_sales):
    print(f"{st}: {val}")

Electronics sales across all stores:
Bengaluru: 850000
Hyderabad: 780000
Chennai: 920000
Mumbai: 1050000
Pune: 690000


---

## Question 7: Clothing Cost Across Stores

Extract the Clothing cost for every store from `cost_data`.

In [ ]:
clothing_index = np.where(categories == "Clothing")[0][0]
clothing_cost = cost_data[:, clothing_index]

print("Clothing cost across all stores:")
for st, val in zip(stores, clothing_cost):
    print(f"{st}: {val}")

Clothing cost across all stores:
Bengaluru: 220000
Hyderabad: 210000
Chennai: 245000
Mumbai: 265000
Pune: 205000


---

## Question 8: Hyderabad Furniture Performance

Find the following for Hyderabad:

* Furniture sales
* Furniture cost
* Furniture profit

In [ ]:
hyderabad_index = np.where(stores == "Hyderabad")[0][0]
furniture_index = np.where(categories == "Furniture")[0][0]

hyderabad_furniture_sales = sales_data[hyderabad_index, furniture_index]
hyderabad_furniture_cost = cost_data[hyderabad_index, furniture_index]
hyderabad_furniture_profit = hyderabad_furniture_sales - hyderabad_furniture_cost

print("Hyderabad Furniture Sales:", hyderabad_furniture_sales)
print("Hyderabad Furniture Cost:", hyderabad_furniture_cost)
print("Hyderabad Furniture Profit:", hyderabad_furniture_profit)

Hyderabad Furniture Sales: 310000
Hyderabad Furniture Cost: 225000
Hyderabad Furniture Profit: 85000


---

## Question 9: Chennai Category Range

Extract Chennai's sales for the following categories:

* Groceries
* Clothing
* Furniture

Use array slicing.

In [ ]:
chennai_index = np.where(stores == "Chennai")[0][0]
chennai_category_range = sales_data[chennai_index, 1:4]

print("Chennai sales for Groceries, Clothing, Furniture:")
for cat, val in zip(categories[1:4], chennai_category_range):
    print(f"{cat}: {val}")

Chennai sales for Groceries, Clothing, Furniture:
Groceries: 390000
Clothing: 340000
Furniture: 260000


---

## Question 10: Selected Store Analysis

Extract sales data for:

* Bengaluru
* Chennai
* Pune

Include only:

* Electronics
* Clothing
* Beauty

Use advanced indexing.

In [ ]:
selected_store_idx = np.where(np.isin(stores, ["Bengaluru", "Chennai", "Pune"]))[0]
selected_category_idx = np.where(np.isin(categories, ["Electronics", "Clothing", "Beauty"]))[0]

selected_store_sales = sales_data[np.ix_(selected_store_idx, selected_category_idx)]

print("Selected stores:", stores[selected_store_idx])
print("Selected categories:", categories[selected_category_idx])
print("\nExtracted sales matrix:")
print(selected_store_sales)

Selected stores: ['Bengaluru' 'Chennai' 'Pune']
Selected categories: ['Electronics' 'Clothing' 'Beauty']

Extracted sales matrix:
[[850000 310000 190000]
 [920000 340000 185000]
 [690000 280000 175000]]


---

# Part 4: Sales Aggregation

## Question 11: Company-Wide Sales

Calculate the total sales revenue generated by all stores and all categories.

In [ ]:
total_company_sales = sales_data.sum()
print("Total company-wide sales revenue:", total_company_sales)

Total company-wide sales revenue: 11675000


---

## Question 12: Store-Wise Sales

Calculate total sales revenue for each store.

Display the result together with the store names.

Expected output format:

```text
Bengaluru: total sales
Hyderabad: total sales
Chennai: total sales
Mumbai: total sales
Pune: total sales
```

In [ ]:
store_wise_sales = sales_data.sum(axis=1)

for st, val in zip(stores, store_wise_sales):
    print(f"{st}: {val}")

Bengaluru: 2270000
Hyderabad: 2295000
Chennai: 2310000
Mumbai: 2820000
Pune: 1980000


---

## Question 13: Category-Wise Sales

Calculate total sales for each product category across all stores.

Display each category name with its total sales.

In [ ]:
category_wise_sales = sales_data.sum(axis=0)

for cat, val in zip(categories, category_wise_sales):
    print(f"{cat}: {val}")

Electronics: 4290000
Groceries: 2185000
Clothing: 1600000
Furniture: 1415000
Beauty: 1010000
Sports: 1175000


---

## Question 14: Average Sales Per Store

Calculate the average sales generated by each store across all six product categories.

In [ ]:
avg_sales_per_store = sales_data.mean(axis=1)

for st, val in zip(stores, avg_sales_per_store):
    print(f"{st}: {val:.2f}")

Bengaluru: 378333.33
Hyderabad: 382500.00
Chennai: 385000.00
Mumbai: 470000.00
Pune: 330000.00


---

## Question 15: Average Sales Per Category

Calculate the average sales of each product category across the five stores.

In [ ]:
avg_sales_per_category = sales_data.mean(axis=0)

for cat, val in zip(categories, avg_sales_per_category):
    print(f"{cat}: {val:.2f}")

Electronics: 858000.00
Groceries: 437000.00
Clothing: 320000.00
Furniture: 283000.00
Beauty: 202000.00
Sports: 235000.00


---

## Question 16: Highest-Selling Store

Identify:

* Index of the store with the highest total sales
* Name of the store
* Total sales generated by that store

Use `np.argmax()`.

In [ ]:
highest_selling_store_idx = np.argmax(store_wise_sales)
highest_selling_store_name = stores[highest_selling_store_idx]
highest_selling_store_value = store_wise_sales[highest_selling_store_idx]

print("Index of highest-selling store:", highest_selling_store_idx)
print("Store name:", highest_selling_store_name)
print("Total sales:", highest_selling_store_value)

Index of highest-selling store: 3
Store name: Mumbai
Total sales: 2820000


---

## Question 17: Lowest-Selling Store

Identify the store with the lowest overall sales.

Use `np.argmin()`.

In [ ]:
lowest_selling_store_idx = np.argmin(store_wise_sales)
lowest_selling_store_name = stores[lowest_selling_store_idx]
lowest_selling_store_value = store_wise_sales[lowest_selling_store_idx]

print("Index of lowest-selling store:", lowest_selling_store_idx)
print("Store name:", lowest_selling_store_name)
print("Total sales:", lowest_selling_store_value)

Index of lowest-selling store: 4
Store name: Pune
Total sales: 1980000


---

## Question 18: Highest-Selling Category

Find the product category that generated the highest total sales across the company.

In [ ]:
highest_selling_category_idx = np.argmax(category_wise_sales)
highest_selling_category_name = categories[highest_selling_category_idx]
highest_selling_category_value = category_wise_sales[highest_selling_category_idx]

print("Highest-selling category:", highest_selling_category_name)
print("Total sales:", highest_selling_category_value)

Highest-selling category: Electronics
Total sales: 4290000


---

## Question 19: Lowest-Selling Category

Find the product category with the lowest total sales across all stores.

In [ ]:
lowest_selling_category_idx = np.argmin(category_wise_sales)
lowest_selling_category_name = categories[lowest_selling_category_idx]
lowest_selling_category_value = category_wise_sales[lowest_selling_category_idx]

print("Lowest-selling category:", lowest_selling_category_name)
print("Total sales:", lowest_selling_category_value)

Lowest-selling category: Beauty
Total sales: 1010000


---

# Part 5: Cost Analysis

## Question 20: Total Company Cost

Calculate the total cost incurred by SmartMart across all stores and product categories.

In [ ]:
total_company_cost = cost_data.sum()
print("Total company-wide cost:", total_company_cost)

Total company-wide cost: 8600000


---

## Question 21: Store-Wise Cost

Calculate the total cost incurred by each store.

In [ ]:
store_wise_cost = cost_data.sum(axis=1)

for st, val in zip(stores, store_wise_cost):
    print(f"{st}: {val}")

Bengaluru: 1665000
Hyderabad: 1685000
Chennai: 1710000
Mumbai: 2065000
Pune: 1475000


---

## Question 22: Category-Wise Cost

Calculate the total cost incurred for each product category across all stores.

In [ ]:
category_wise_cost = cost_data.sum(axis=0)

for cat, val in zip(categories, category_wise_cost):
    print(f"{cat}: {val}")

Electronics: 3250000
Groceries: 1660000
Clothing: 1145000
Furniture: 1045000
Beauty: 665000
Sports: 835000


---

## Question 23: Most Expensive Store

Identify the store with the highest total operating cost.

In [ ]:
most_expensive_store_idx = np.argmax(store_wise_cost)
most_expensive_store_name = stores[most_expensive_store_idx]

print("Most expensive store:", most_expensive_store_name)
print("Total cost:", store_wise_cost[most_expensive_store_idx])

Most expensive store: Mumbai
Total cost: 2065000


---

## Question 24: Lowest-Cost Store

Identify the store with the lowest total cost.

In [ ]:
lowest_cost_store_idx = np.argmin(store_wise_cost)
lowest_cost_store_name = stores[lowest_cost_store_idx]

print("Lowest-cost store:", lowest_cost_store_name)
print("Total cost:", store_wise_cost[lowest_cost_store_idx])

Lowest-cost store: Pune
Total cost: 1475000


---

## Question 25: Most Expensive Product Category

Determine which product category has the highest total cost across all stores.

In [ ]:
most_expensive_category_idx = np.argmax(category_wise_cost)
most_expensive_category_name = categories[most_expensive_category_idx]

print("Most expensive category:", most_expensive_category_name)
print("Total cost:", category_wise_cost[most_expensive_category_idx])

Most expensive category: Electronics
Total cost: 3250000


---

# Part 6: Profit Analysis Using Two Arrays

## Question 26: Profit Matrix

Calculate the profit generated for every store and product category.

Use:

```python
profit_data = sales_data - cost_data
```

Display the complete profit matrix.

In [ ]:
profit_data = sales_data - cost_data
print("Profit matrix (stores x categories):")
print(profit_data)

Profit matrix (stores x categories):
[[210000 105000  90000  70000  65000  65000]
 [190000 110000  85000  85000  70000  70000]
 [220000  95000  95000  65000  65000  60000]
 [260000 120000 110000  95000  85000  85000]
 [160000  95000  75000  55000  60000  60000]]


---

## Question 27: Store-Wise Profit

Calculate the total profit generated by each store.

In [ ]:
store_wise_profit = profit_data.sum(axis=1)

for st, val in zip(stores, store_wise_profit):
    print(f"{st}: {val}")

Bengaluru: 605000
Hyderabad: 610000
Chennai: 600000
Mumbai: 755000
Pune: 505000


---

## Question 28: Category-Wise Profit

Calculate the total profit generated by each product category.

In [ ]:
category_wise_profit = profit_data.sum(axis=0)

for cat, val in zip(categories, category_wise_profit):
    print(f"{cat}: {val}")

Electronics: 1040000
Groceries: 525000
Clothing: 455000
Furniture: 370000
Beauty: 345000
Sports: 340000


---

## Question 29: Most Profitable Store

Identify the store that generated the highest total profit.

Display:

* Store name
* Total sales
* Total cost
* Total profit

In [ ]:
most_profitable_store_idx = np.argmax(store_wise_profit)
most_profitable_store_name = stores[most_profitable_store_idx]

print("Store name:", most_profitable_store_name)
print("Total sales:", store_wise_sales[most_profitable_store_idx])
print("Total cost:", store_wise_cost[most_profitable_store_idx])
print("Total profit:", store_wise_profit[most_profitable_store_idx])

Store name: Mumbai
Total sales: 2820000
Total cost: 2065000
Total profit: 755000


---

## Question 30: Least Profitable Store

Identify the store that generated the lowest total profit.

In [ ]:
least_profitable_store_idx = np.argmin(store_wise_profit)
least_profitable_store_name = stores[least_profitable_store_idx]

print("Store name:", least_profitable_store_name)
print("Total sales:", store_wise_sales[least_profitable_store_idx])
print("Total cost:", store_wise_cost[least_profitable_store_idx])
print("Total profit:", store_wise_profit[least_profitable_store_idx])

Store name: Pune
Total sales: 1980000
Total cost: 1475000
Total profit: 505000


---

## Question 31: Most Profitable Category

Find the category with the highest total profit across all stores.

In [ ]:
most_profitable_category_idx = np.argmax(category_wise_profit)
most_profitable_category_name = categories[most_profitable_category_idx]

print("Most profitable category:", most_profitable_category_name)
print("Total profit:", category_wise_profit[most_profitable_category_idx])

Most profitable category: Electronics
Total profit: 1040000


---

## Question 32: Least Profitable Category

Find the category with the lowest total profit.

In [ ]:
least_profitable_category_idx = np.argmin(category_wise_profit)
least_profitable_category_name = categories[least_profitable_category_idx]

print("Least profitable category:", least_profitable_category_name)
print("Total profit:", category_wise_profit[least_profitable_category_idx])

Least profitable category: Sports
Total profit: 340000


---

## Question 33: Maximum Individual Profit

Find the highest profit generated by any single store-category combination.

Display:

* Profit amount
* Store name
* Product category

Use `np.unravel_index()`.

In [ ]:
max_profit_flat_idx = np.argmax(profit_data)
max_profit_row, max_profit_col = np.unravel_index(max_profit_flat_idx, profit_data.shape)

print("Maximum individual profit:", profit_data[max_profit_row, max_profit_col])
print("Store:", stores[max_profit_row])
print("Category:", categories[max_profit_col])

Maximum individual profit: 260000
Store: Mumbai
Category: Electronics


---

## Question 34: Minimum Individual Profit

Find the smallest profit generated by a store-category combination.

Display the corresponding store and category.

In [ ]:
min_profit_flat_idx = np.argmin(profit_data)
min_profit_row, min_profit_col = np.unravel_index(min_profit_flat_idx, profit_data.shape)

print("Minimum individual profit:", profit_data[min_profit_row, min_profit_col])
print("Store:", stores[min_profit_row])
print("Category:", categories[min_profit_col])

Minimum individual profit:

 55000
Store: Pune
Category: Furniture


---

# Part 7: Profit Margin Analysis

## Question 35: Profit Margin Matrix

Calculate the profit margin percentage for every store-category combination.

Use:

```python
profit_margin = (profit_data / sales_data) * 100
```

Round the result to two decimal places.

In [ ]:
profit_margin = np.round((profit_data / sales_data) * 100, 2)
print("Profit margin matrix (%):")
print(profit_margin)

Profit margin matrix (%):
[[24.71 25.   29.03 25.45 34.21 28.89]
 [24.36 23.91 28.81 27.42 33.33 29.17]
 [23.91 24.36 27.94 25.   35.14 27.91]
 [24.76 23.53 29.33 27.54 34.   29.31]
 [23.19 23.46 26.79 24.44 34.29 29.27]]


---

## Question 36: Store-Wise Profit Margin

Calculate the overall profit margin percentage for each store using:

```text
Store Profit Margin =
Total Store Profit / Total Store Sales × 100
```

In [ ]:
store_profit_margin = np.round((store_wise_profit / store_wise_sales) * 100, 2)

for st, val in zip(stores, store_profit_margin):
    print(f"{st}: {val}%")

Bengaluru: 26.65%
Hyderabad: 26.58%
Chennai: 25.97%
Mumbai: 26.77%
Pune: 25.51%


---

## Question 37: Category-Wise Profit Margin

Calculate the overall profit margin percentage for each product category.

In [ ]:
category_profit_margin = np.round((category_wise_profit / category_wise_sales) * 100, 2)

for cat, val in zip(categories, category_profit_margin):
    print(f"{cat}: {val}%")

Electronics: 24.24%
Groceries: 24.03%
Clothing: 28.44%
Furniture: 26.15%
Beauty: 34.16%
Sports: 28.94%


---

## Question 38: Best Store by Profit Margin

Identify the store with the highest overall profit margin.

In [ ]:
best_margin_store_idx = np.argmax(store_profit_margin)
print("Store with highest profit margin:", stores[best_margin_store_idx])
print("Profit margin:", store_profit_margin[best_margin_store_idx], "%")

Store with highest profit margin: Mumbai
Profit margin: 26.77 %


---

## Question 39: Best Category by Profit Margin

Identify the product category with the highest profit margin.

In [ ]:
best_margin_category_idx = np.argmax(category_profit_margin)
print("Category with highest profit margin:", categories[best_margin_category_idx])
print("Profit margin:", category_profit_margin[best_margin_category_idx], "%")

Category with highest profit margin: Beauty
Profit margin: 34.16 %


---

## Question 40: Low-Margin Transactions

Management considers a profit margin below 22% to be weak.

Find all store-category combinations where the profit margin is below 22%.

Display:

* Store
* Category
* Sales
* Cost
* Profit
* Profit margin

Use Boolean indexing.

In [ ]:
low_margin_mask = profit_margin < 22
low_margin_store_idx, low_margin_cat_idx = np.where(low_margin_mask)

print("Low-margin (< 22%) store-category combinations:")
for r, c in zip(low_margin_store_idx, low_margin_cat_idx):
    print(f"Store: {stores[r]}, Category: {categories[c]}, "
          f"Sales: {sales_data[r, c]}, Cost: {cost_data[r, c]}, "
          f"Profit: {profit_data[r, c]}, Margin: {profit_margin[r, c]}%")

Low-margin (< 22%) store-category combinations:


---

# Part 8: Boolean Filtering

## Question 41: High Sales Values

Find all sales values greater than ₹5,00,000.

In [ ]:
high_sales_values = sales_data[sales_data > 500000]
print("Sales values greater than 500000:")
print(high_sales_values)

Sales values greater than 500000:
[ 850000  780000  920000 1050000  510000  690000]


---

## Question 42: Low Sales Values

Find all sales values below ₹2,00,000.

In [ ]:
low_sales_values = sales_data[sales_data < 200000]
print("Sales values below 200000:")
print(low_sales_values)

Sales values below 200000:
[190000 185000 175000]


---

## Question 43: High Profit Values

Find all store-category combinations where profit is greater than ₹1,50,000.

In [ ]:
high_profit_values = profit_data[profit_data > 150000]
print("Profit values greater than 150000:")
print(high_profit_values)

Profit values greater than 150000:
[210000 190000 220000 260000 160000]


---

## Question 44: Moderate Sales Range

Find sales values between ₹2,50,000 and ₹5,00,000.

Use `np.logical_and()`.

In [ ]:
moderate_sales_mask = np.logical_and(sales_data >= 250000, sales_data <= 500000)
moderate_sales = sales_data[moderate_sales_mask]
print("Sales values between 250000 and 500000:")
print(moderate_sales)

Sales values between 250000 and 500000:
[420000 310000 275000 460000 295000 310000 390000 340000 260000 375000
 345000 250000 290000 405000 280000]


---

## Question 45: Premium Performance

A store-category combination is considered premium when:

* Sales are greater than ₹5,00,000
* Profit is greater than ₹1,50,000

Find all premium-performing combinations.

In [ ]:
premium_mask = np.logical_and(sales_data > 500000, profit_data > 150000)
premium_store_idx, premium_cat_idx = np.where(premium_mask)

print("Premium-performing combinations:")
for r, c in zip(premium_store_idx, premium_cat_idx):
    print(f"Store: {stores[r]}, Category: {categories[c]}, "
          f"Sales: {sales_data[r, c]}, Profit: {profit_data[r, c]}")

Premium-performing combinations:


Store: Bengaluru, Category: Electronics, Sales: 850000, Profit: 210000
Store: Hyderabad, Category: Electronics, Sales: 780000, Profit: 190000
Store: Chennai, Category: Electronics, Sales: 920000, Profit: 220000
Store: Mumbai, Category: Electronics, Sales: 1050000, Profit: 260000
Store: Pune, Category: Electronics, Sales: 690000, Profit: 160000


---

## Question 46: Underperforming Combinations

A store-category combination is considered underperforming when:

* Sales are below ₹2,50,000
* Profit margin is below 25%

Find all such combinations.

In [ ]:
underperforming_mask = np.logical_and(sales_data < 250000, profit_margin < 25)
under_store_idx, under_cat_idx = np.where(underperforming_mask)

print("Underperforming combinations:")
for r, c in zip(under_store_idx, under_cat_idx):
    print(f"Store: {stores[r]}, Category: {categories[c]}, "
          f"Sales: {sales_data[r, c]}, Margin: {profit_margin[r, c]}%")

Underperforming combinations:
Store: Pune, Category: Furniture, Sales: 225000, Margin: 24.44%


---

## Question 47: Category Performance Classification

Create a new array that classifies every sales value as:

* `"High"` when sales are at least ₹5,00,000
* `"Medium"` when sales are between ₹2,50,000 and ₹4,99,999
* `"Low"` when sales are below ₹2,50,000

Use `np.where()` or `np.select()`.

In [ ]:
sales_classification = np.select(
    [sales_data >= 500000, sales_data >= 250000],
    ["High", "Medium"],
    default="Low"
)

print("Sales classification matrix:")
print(sales_classification)

Sales classification matrix:
[['High' 'Medium' 'Medium' 'Medium' 'Low' 'Low']
 ['High' 'Medium' 'Medium' 'Medium' 'Low' 'Low']
 ['High' 'Medium' 'Medium' 'Medium' 'Low' 'Low']
 ['High' 'High' 'Medium' 'Medium' 'Medium' 'Medium']
 ['High' 'Medium' 'Medium' 'Low' 'Low' 'Low']]


---

# Part 9: Target-Based Analysis

Use the following category-wise monthly sales targets:

```python
sales_targets = np.array([
    800000,  # Electronics
    450000,  # Groceries
    320000,  # Clothing
    300000,  # Furniture
    200000,  # Beauty
    230000   # Sports
])
```

## Question 48: Target Comparison

Compare every store's category sales against `sales_targets`.

Use NumPy broadcasting.

In [ ]:
sales_targets = np.array([
    800000,  # Electronics
    450000,  # Groceries
    320000,  # Clothing
    300000,  # Furniture
    200000,  # Beauty
    230000   # Sports
])

target_comparison = sales_data - sales_targets
print("Sales minus target, for every store-category combination:")
print(target_comparison)

Sales minus target, for every store-category combination:
[[  50000  -30000  -10000  -25000  -10000   -5000]
 [ -20000   10000  -25000   10000   10000   10000]
 [ 120000  -60000   20000  -40000  -15000  -15000]
 [ 250000   60000   55000   45000   50000   60000]
 [-110000  -45000  -40000  -75000  -25000  -25000]]


---

## Question 49: Target Achievement Matrix

Create a Boolean array where:

* `True` means the store achieved the category target
* `False` means the store failed to achieve the target

In [ ]:
target_achieved = sales_data >= sales_targets
print("Target achievement matrix:")
print(target_achieved)

Target achievement matrix:
[[ True False False False False False]
 [False  True False  True  True  True]
 [ True False  True False False False]
 [ True  True  True  True  True  True]
 [False False False False False False]]


---

## Question 50: Target Achievement Count

Calculate the number of categories in which each store achieved its target.

In [ ]:
achievement_count = target_achieved.sum(axis=1)

for st, val in zip(stores, achievement_count):
    print(f"{st}: {val} out of {len(categories)} targets achieved")

Bengaluru: 1 out of 6 targets achieved
Hyderabad: 4 out of 6 targets achieved
Chennai: 2 out of 6 targets achieved
Mumbai: 6 out of 6 targets achieved
Pune: 0 out of 6 targets achieved


---

## Question 51: Best Target-Performing Store

Find the store that achieved the highest number of category targets.

In [ ]:
best_target_store_idx = np.argmax(achievement_count)
print("Best target-performing store:", stores[best_target_store_idx])
print("Targets achieved:", achievement_count[best_target_store_idx])

Best target-performing store: Mumbai
Targets achieved: 6


---

## Question 52: Target Achievement by Category

Calculate how many stores achieved the target for each product category.

In [ ]:
achievement_by_category = target_achieved.sum(axis=0)

for cat, val in zip(categories, achievement_by_category):
    print(f"{cat}: {val} out of {len(stores)} stores achieved target")

Electronics: 3 out of 5 stores achieved target
Groceries: 2 out of 5 stores achieved target
Clothing: 2 out of 5 stores achieved target
Furniture: 2 out of 5 stores achieved target
Beauty: 2 out of 5 stores achieved target
Sports: 2 out of 5 stores achieved target


---

## Question 53: Sales Target Variance

Calculate the difference between actual sales and target sales for every store.

Use:

```text
Variance = Actual Sales - Target Sales
```

In [ ]:
target_variance = sales_data - sales_targets
print("Sales target variance matrix:")
print(target_variance)

Sales target variance matrix:
[[  50000  -30000  -10000  -25000  -10000   -5000]
 [ -20000   10000  -25000   10000   10000   10000]
 [ 120000  -60000   20000  -40000  -15000  -15000]
 [ 250000   60000   55000   45000   50000   60000]
 [-110000  -45000  -40000  -75000  -25000  -25000]]


---

## Question 54: Percentage Above or Below Target

Calculate the percentage variance from the target using:

```text
Percentage Variance =
(Actual Sales - Target Sales) / Target Sales × 100
```

In [ ]:
pct_target_variance = np.round((target_variance / sales_targets) * 100, 2)
print("Percentage variance from target:")
print(pct_target_variance)

Percentage variance from target:
[[  6.25  -6.67  -3.12  -8.33  -5.    -2.17]
 [ -2.5    2.22  -7.81   3.33   5.     4.35]
 [ 15.   -13.33   6.25 -13.33  -7.5   -6.52]
 [ 31.25  13.33  17.19  15.    25.    26.09]
 [-13.75 -10.   -12.5  -25.   -12.5  -10.87]]


---

## Question 55: Stores Missing Electronics Target

Identify stores that failed to achieve the Electronics sales target.

In [ ]:
electronics_col = np.where(categories == "Electronics")[0][0]
electronics_missed = stores[~target_achieved[:, electronics_col]]

print("Stores that missed the Electronics target:")
print(electronics_missed)

Stores that missed the Electronics target:
['Hyderabad' 'Pune']


---

## Question 56: Stores Achieving All Targets

Check whether any store achieved its target in all six product categories.

Use `np.all()`.

In [ ]:
all_targets_per_store = np.all(target_achieved, axis=1)
stores_all_targets = stores[all_targets_per_store]

print("Did every-target achievement happen for any store?", all_targets_per_store.any())
print("Stores that achieved ALL targets:", stores_all_targets if stores_all_targets.size else "None")

Did every-target achievement happen for any store? True
Stores that achieved ALL targets: ['Mumbai']


---

## Question 57: Stores Achieving At Least One Target

Identify stores that achieved at least one category target.

Use `np.any()`.

In [ ]:
at_least_one_target = np.any(target_achieved, axis=1)
stores_at_least_one = stores[at_least_one_target]

print("Stores that achieved at least one target:")
print(stores_at_least_one)

Stores that achieved at least one target:


['Bengaluru' 'Hyderabad' 'Chennai' 'Mumbai']


---

# Part 10: Discount Scenario

SmartMart plans to offer category-wise discounts:

```python
discount_rates = np.array([
    0.10,  # Electronics: 10%
    0.05,  # Groceries: 5%
    0.15,  # Clothing: 15%
    0.12,  # Furniture: 12%
    0.08,  # Beauty: 8%
    0.10   # Sports: 10%
])
```

## Question 58: Discount Amount Matrix

Calculate the discount amount for each store-category combination.

Use broadcasting.

In [ ]:
discount_rates = np.array([
    0.10,  # Electronics: 10%
    0.05,  # Groceries: 5%
    0.15,  # Clothing: 15%
    0.12,  # Furniture: 12%
    0.08,  # Beauty: 8%
    0.10   # Sports: 10%
])

discount_amount = sales_data * discount_rates
print("Discount amount matrix:")
print(discount_amount)

Discount amount matrix:
[[ 85000.  21000.  46500.  33000.  15200.  22500.]
 [ 78000.  23000.  44250.  37200.  16800.  24000.]
 [ 92000.  19500.  51000.  31200.  14800.  21500.]
 [105000.  25500.  56250.  41400.  20000.  29000.]
 [ 69000.  20250.  42000.  27000.  14000.  20500.]]


---

## Question 59: Sales After Discount

Calculate the net sales value after applying the category-wise discounts.

In [ ]:
sales_after_discount = sales_data - discount_amount
print("Sales after discount:")
print(sales_after_discount)

Sales after discount:
[[765000. 399000. 263500. 242000. 174800. 202500.]
 [702000. 437000. 250750. 272800. 193200. 216000.]
 [828000. 370500. 289000. 228800. 170200. 193500.]
 [945000. 484500. 318750. 303600. 230000. 261000.]
 [621000. 384750. 238000. 198000. 161000. 184500.]]


---

## Question 60: Profit After Discount

Assuming costs remain unchanged, calculate the new profit after the discounts are applied.

In [ ]:
profit_after_discount = sales_after_discount - cost_data
print("Profit after discount:")
print(profit_after_discount)

Profit after discount:


[[125000.  84000.  43500.  37000.  49800.  42500.]
 [112000.  87000.  40750.  47800.  53200.  46000.]
 [128000.  75500.  44000.  33800.  50200.  38500.]
 [155000.  94500.  53750.  53600.  65000.  56000.]
 [ 91000.  74750.  33000.  28000.  46000.  39500.]]


---

## Question 61: Profit Margin After Discount

Calculate the new profit margin after applying discounts.

In [ ]:
margin_after_discount = np.round((profit_after_discount / sales_after_discount) * 100, 2)
print("Profit margin after discount (%):")
print(margin_after_discount)

Profit margin after discount (%):
[[16.34 21.05 16.51 15.29 28.49 20.99]
 [15.95 19.91 16.25 17.52 27.54 21.3 ]
 [15.46 20.38 15.22 14.77 29.49 19.9 ]
 [16.4  19.5  16.86 17.65 28.26 21.46]
 [14.65 19.43 13.87 14.14 28.57 21.41]]


---

## Question 62: Loss-Making Transactions

Check whether any store-category combination becomes loss-making after discounts.

Display the affected store and category names.

In [ ]:
loss_after_discount_mask = profit_after_discount < 0
loss_store_idx, loss_cat_idx = np.where(loss_after_discount_mask)

if loss_store_idx.size:
    print("Store-category combinations that become loss-making after discount:")
    for r, c in zip(loss_store_idx, loss_cat_idx):
        print(f"Store: {stores[r]}, Category: {categories[c]}, "
              f"Profit after discount: {profit_after_discount[r, c]}")
else:
    print("No store-category combination becomes loss-making after the proposed discounts.")

No store-category combination becomes loss-making after the proposed discounts.


---

## Question 63: Discount Impact

Calculate how much total revenue the company loses because of the proposed discounts.

In [ ]:
total_discount_impact = discount_amount.sum()
print("Total revenue lost due to discounts:", total_discount_impact)

Total revenue lost due to discounts: 1146350.0


---

## Question 64: Store Most Affected by Discounts

Identify the store that loses the highest amount of revenue due to discounts.

In [ ]:
store_wise_discount = discount_amount.sum(axis=1)
store_most_affected_idx = np.argmax(store_wise_discount)

print("Store most affected by discounts:", stores[store_most_affected_idx])
print("Revenue lost:", store_wise_discount[store_most_affected_idx])

Store most affected by discounts: Mumbai
Revenue lost: 277150.0


---

# Part 11: Tax and Price Adjustment

The government introduces category-wise tax rates:

```python
tax_rates = np.array([
    0.18,
    0.05,
    0.12,
    0.18,
    0.18,
    0.12
])
```

## Question 65: Tax Collection Matrix

Calculate the tax collected on sales for each store and category.

In [ ]:
tax_rates = np.array([
    0.18,  # Electronics
    0.05,  # Groceries
    0.12,  # Clothing
    0.18,  # Furniture
    0.18,  # Beauty
    0.12   # Sports
])

tax_matrix = sales_data * tax_rates
print("Tax collection matrix:")
print(tax_matrix)

Tax collection matrix:
[[153000.  21000.  37200.  49500.  34200.  27000.]
 [140400.  23000.  35400.  55800.  37800.  28800.]
 [165600.  19500.  40800.  46800.  33300.  25800.]
 [189000.  25500.  45000.  62100.  45000.  34800.]
 [124200.  20250.  33600.  40500.  31500.  24600.]]


---

## Question 66: Sales Including Tax

Calculate the final customer billing value after adding tax.

In [ ]:
sales_incl_tax = sales_data + tax_matrix
print("Sales including tax:")
print(sales_incl_tax)

Sales including tax:
[[1003000.  441000.  347200.  324500.  224200.  252000.]
 [ 920400.  483000.  330400.  365800.  247800.  268800.]
 [1085600.  409500.  380800.  306800.  218300.  240800.]
 [1239000.  535500.  420000.  407100.  295000.  324800.]
 [ 814200.  425250.  313600.  265500.  206500.  229600.]]


---

## Question 67: Store-Wise Tax Collection

Calculate the total tax collected by each store.

In [ ]:
store_wise_tax = tax_matrix.sum(axis=1)

for st, val in zip(stores, store_wise_tax):
    print(f"{st}: {val:.2f}")

Bengaluru: 321900.00
Hyderabad: 321200.00
Chennai: 331800.00
Mumbai: 401400.00
Pune: 274650.00


---

## Question 68: Category-Wise Tax Collection

Calculate total tax collected for each category.

In [ ]:
category_wise_tax = tax_matrix.sum(axis=0)

for cat, val in zip(categories, category_wise_tax):
    print(f"{cat}: {val:.2f}")

Electronics: 772200.00
Groceries: 109250.00
Clothing: 192000.00
Furniture: 254700.00
Beauty: 181800.00
Sports: 141000.00


---

## Question 69: Highest Tax-Contributing Store

Identify the store that contributes the highest tax amount.

In [ ]:
highest_tax_store_idx = np.argmax(store_wise_tax)
print("Highest tax-contributing store:", stores[highest_tax_store_idx])
print("Tax contributed:", store_wise_tax[highest_tax_store_idx])

Highest tax-contributing store: Mumbai
Tax contributed: 401400.0


---

# Part 12: Array Modification

## Question 70: Updated Pune Electronics Sales

The Pune store reports that its Electronics sales were entered incorrectly.

Update Pune Electronics sales from ₹6,90,000 to ₹7,40,000.

Make the change in a copied array without modifying the original array.

In [ ]:
pune_idx = np.where(stores == "Pune")[0][0]
electronics_idx = np.where(categories == "Electronics")[0][0]

updated_sales = sales_data.copy()
updated_sales[pune_idx, electronics_idx] = 740000

print("Original Pune Electronics sales:", sales_data[pune_idx, electronics_idx])
print("Updated Pune Electronics sales:", updated_sales[pune_idx, electronics_idx])

Original Pune Electronics sales: 690000
Updated Pune Electronics sales: 740000


---

## Question 71: Revised Beauty Cost

Due to supplier negotiations, Beauty product costs are reduced by 8% for all stores.

Create an updated cost array.

In [ ]:
beauty_idx = np.where(categories == "Beauty")[0][0]

revised_beauty_cost = cost_data.copy()
revised_beauty_cost[:, beauty_idx] = revised_beauty_cost[:, beauty_idx] * 0.92

print("Original Beauty cost:", cost_data[:, beauty_idx])
print("Revised Beauty cost (8% reduction):", revised_beauty_cost[:, beauty_idx])

Original Beauty cost: [125000 140000 120000 165000 115000]
Revised Beauty cost (8% reduction): [115000 128800 110400 151800 105800]


---

## Question 72: Furniture Cost Increase

Furniture costs increase by 12% for all stores.

Update only the Furniture column.

In [ ]:
furniture_idx = np.where(categories == "Furniture")[0][0]

furniture_cost_increased = cost_data.copy()
furniture_cost_increased[:, furniture_idx] = furniture_cost_increased[:, furniture_idx] * 1.12

print("Original Furniture cost:", cost_data[:, furniture_idx])
print("Increased Furniture cost (12% increase):", furniture_cost_increased[:, furniture_idx])

Original Furniture cost: [205000 225000 195000 250000 170000]
Increased Furniture cost (12% increase): [229600 252000 218400 280000 190400]


---

## Question 73: Store Closure Simulation

Assume the Pune store is temporarily closed.

Remove Pune's sales and cost data from the analysis using `np.delete()`.

In [ ]:
sales_without_pune = np.delete(sales_data, pune_idx, axis=0)
cost_without_pune = np.delete(cost_data, pune_idx, axis=0)
stores_without_pune = np.delete(stores, pune_idx)

print("Stores after removing Pune:", stores_without_pune)
print("\nSales data after removing Pune:")
print(sales_without_pune)
print("\nCost data after removing Pune:")
print(cost_without_pune)

Stores after removing Pune: ['Bengaluru' 'Hyderabad' 'Chennai' 'Mumbai']

Sales data after removing Pune:
[[ 850000  420000  310000  275000  190000  225000]
 [ 780000  460000  295000  310000  210000  240000]
 [ 920000  390000  340000  260000  185000  215000]
 [1050000  510000  375000  345000  250000  290000]]

Cost data after removing Pune:
[[640000 315000 220000 205000 125000 160000]
 [590000 350000 210000 225000 140000 170000]
 [700000 295000 245000 195000 120000 155000]
 [790000 390000 265000 250000 165000 205000]]


---

## Question 74: Adding a New Store

A new Delhi store is opened with the following data:

```python
delhi_sales = np.array([
    880000, 475000, 330000, 290000, 215000, 250000
])

delhi_cost = np.array([
    665000, 360000, 235000, 215000, 145000, 180000
])
```

Add Delhi to both arrays using `np.vstack()`.

Also add `"Delhi"` to the stores array.

In [ ]:
delhi_sales = np.array([880000, 475000, 330000, 290000, 215000, 250000])
delhi_cost = np.array([665000, 360000, 235000, 215000, 145000, 180000])

sales_with_delhi = np.vstack([sales_data, delhi_sales])
cost_with_delhi = np.vstack([cost_data, delhi_cost])
stores_with_delhi = np.append(stores, "Delhi")

print("Stores after adding Delhi:", stores_with_delhi)
print("\nSales data after adding Delhi:")
print(sales_with_delhi)
print("\nCost data after adding Delhi:")
print(cost_with_delhi)

Stores after adding Delhi: ['Bengaluru' 'Hyderabad' 'Chennai' 'Mumbai' 'Pune' 'Delhi']

Sales data after adding Delhi:
[[ 850000  420000  310000  275000  190000  225000]
 [ 780000  460000  295000  310000  210000  240000]
 [ 920000  390000  340000  260000  185000  215000]
 [1050000  510000  375000  345000  250000  290000]
 [ 690000  405000  280000  225000  175000  205000]
 [ 880000  475000  330000  290000  215000  250000]]

Cost data after adding Delhi:
[[640000 315000 220000 205000 125000 160000]
 [590000 350000 210000 225000 140000 170000]
 [700000 295000 245000 195000 120000 155000]
 [790000 390000 265000 250000 165000 205000]
 [530000 310000 205000 170000 115000 145000]
 [665000 360000 235000 215000 145000 180000]]


---

## Question 75: Adding a New Product Category

SmartMart starts selling Home Appliances.

```python
home_appliance_sales = np.array([
    360000,
    325000,
    390000,
    420000,
    295000
])

home_appliance_cost = np.array([
    265000,
    240000,
    285000,
    305000,
    225000
])
```

Add the new category as a column using `np.column_stack()` or `np.hstack()`.

In [ ]:
home_appliance_sales = np.array([360000, 325000, 390000, 420000, 295000])
home_appliance_cost = np.array([265000, 240000, 285000, 305000, 225000])

sales_with_appliances = np.column_stack([sales_data, home_appliance_sales])
cost_with_appliances = np.column_stack([cost_data, home_appliance_cost])
categories_with_appliances = np.append(categories, "Home Appliances")

print("Categories after adding Home Appliances:", categories_with_appliances)
print("\nSales data after adding Home Appliances:")
print(sales_with_appliances)
print("\nCost data after adding Home Appliances:")
print(cost_with_appliances)

Categories after adding Home Appliances: ['Electronics' 'Groceries' 'Clothing' 'Furniture' 'Beauty' 'Sports'
 'Home Appliances']

Sales data after adding Home Appliances:
[[ 850000  420000  310000  275000  190000  225000  360000]
 [ 780000  460000  295000  310000  210000  240000  325000]
 [ 920000  390000  340000  260000  185000  215000  390000]
 [1050000  510000  375000  345000  250000  290000  420000]
 [ 690000  405000  280000  225000  175000  205000  295000]]

Cost data after adding Home Appliances:
[[640000 315000 220000 205000 125000 160000 265000]
 [590000 350000 210000 225000 140000 170000 240000]
 [700000 295000 245000 195000 120000 155000 285000]
 [790000 390000 265000 250000 165000 205000 305000]
 [530000 310000 205000 170000 115000 145000 225000]]


---

# Part 13: Sorting and Ranking

## Question 76: Sort Stores by Sales

Sort stores in ascending order based on total sales.

Display store names and sales totals.

In [ ]:
ascending_sales_order = np.argsort(store_wise_sales)

print("Stores sorted by total sales (ascending):")
for idx in ascending_sales_order:
    print(f"{stores[idx]}: {store_wise_sales[idx]}")

Stores sorted by total sales (ascending):
Pune: 1980000
Bengaluru: 2270000
Hyderabad: 2295000
Chennai: 2310000
Mumbai: 2820000


---

## Question 77: Rank Stores by Profit

Rank stores from highest to lowest total profit.

Use `np.argsort()`.

In [ ]:
descending_profit_order = np.argsort(store_wise_profit)[::-1]

print("Stores ranked by total profit (highest to lowest):")
for rank, idx in enumerate(descending_profit_order, start=1):
    print(f"{rank}. {stores[idx]}: {store_wise_profit[idx]}")

Stores ranked by total profit (highest to lowest):
1. Mumbai: 755000
2. Hyderabad: 610000
3. Bengaluru: 605000
4. Chennai: 600000
5. Pune: 505000


---

## Question 78: Sort Categories by Revenue

Sort product categories from highest to lowest total revenue.

In [ ]:
descending_revenue_order = np.argsort(category_wise_sales)[::-1]

print("Categories sorted by total revenue (highest to lowest):")
for idx in descending_revenue_order:
    print(f"{categories[idx]}: {category_wise_sales[idx]}")

Categories sorted by total revenue (highest to lowest):
Electronics: 4290000
Groceries: 2185000
Clothing: 1600000
Furniture: 1415000
Sports: 1175000
Beauty: 1010000


---

## Question 79: Top Three Store-Category Sales

Find the three highest individual sales values in the complete sales matrix.

Display their:

* Sales value
* Store name
* Category name

In [ ]:
flat_sales = sales_data.flatten()
top3_flat_idx = np.argsort(flat_sales)[::-1][:3]
top3_coords = [np.unravel_index(idx, sales_data.shape) for idx in top3_flat_idx]

print("Top 3 individual store-category sales values:")
for r, c in top3_coords:
    print(f"Sales: {sales_data[r, c]}, Store: {stores[r]}, Category: {categories[c]}")

Top 3 individual store-category sales values:
Sales: 1050000, Store: Mumbai, Category: Electronics
Sales: 920000, Store: Chennai, Category: Electronics
Sales: 850000, Store: Bengaluru, Category: Electronics


---

## Question 80: Bottom Three Profits

Find the three lowest profit values in the profit matrix.

In [ ]:
flat_profit = profit_data.flatten()
bottom3_flat_idx = np.argsort(flat_profit)[:3]
bottom3_coords = [np.unravel_index(idx, profit_data.shape) for idx in bottom3_flat_idx]

print("Bottom 3 individual store-category profit values:")
for r, c in bottom3_coords:
    print(f"Profit: {profit_data[r, c]}, Store: {stores[r]}, Category: {categories[c]}")

Bottom 3 individual store-category profit values:
Profit: 55000, Store: Pune, Category: Furniture
Profit: 60000, Store: Pune, Category: Sports
Profit: 60000, Store: Chennai, Category: Sports


---

# Part 14: Statistical Analysis

## Question 81: Sales Statistics

Calculate the following for the complete sales array:

* Minimum
* Maximum
* Mean
* Median
* Standard deviation
* Variance
* 25th percentile
* 50th percentile
* 75th percentile

In [ ]:
sales_min = sales_data.min()
sales_max = sales_data.max()
sales_mean = sales_data.mean()
sales_median = np.median(sales_data)
sales_std = sales_data.std()
sales_var = sales_data.var()
sales_p25 = np.percentile(sales_data, 25)
sales_p50 = np.percentile(sales_data, 50)
sales_p75 = np.percentile(sales_data, 75)

print("Minimum:", sales_min)
print("Maximum:", sales_max)
print("Mean:", round(sales_mean, 2))
print("Median:", sales_median)
print("Standard Deviation:", round(sales_std, 2))
print("Variance:", round(sales_var, 2))
print("25th Percentile:", sales_p25)
print("50th Percentile:", sales_p50)
print("75th Percentile:", sales_p75)

Minimum:

 175000
Maximum: 1050000
Mean: 389166.67
Median: 302500.0
Standard Deviation: 230322.54
Variance: 53048472222.22
25th Percentile: 228750.0
50th Percentile: 302500.0
75th Percentile: 416250.0


---

## Question 82: Store-Wise Standard Deviation

Calculate the sales standard deviation for every store.

Use this value to determine which store has the most variation between product categories.

In [ ]:
store_std = sales_data.std(axis=1)

for st, val in zip(stores, store_std):
    print(f"{st}: {val:.2f}")

most_variable_store_idx = np.argmax(store_std)
print("\nStore with the most variation across categories:", stores[most_variable_store_idx])

Bengaluru: 223059.53
Hyderabad: 194502.57
Chennai: 249265.59
Mumbai: 271860.87
Pune: 177200.45

Store with the most variation across categories: Mumbai


---

## Question 83: Category-Wise Standard Deviation

Determine which product category has the highest variation in sales across stores.

In [ ]:
category_std = sales_data.std(axis=0)

for cat, val in zip(categories, category_std):
    print(f"{cat}: {val:.2f}")

most_variable_category_idx = np.argmax(category_std)
print("\nCategory with the highest variation across stores:", categories[most_variable_category_idx])

Electronics: 122539.79
Groceries: 43312.82
Clothing: 33911.65
Furniture: 41303.75
Beauty: 26570.66
Sports: 29832.87

Category with the highest variation across stores: Electronics


---

## Question 84: Sales Above Company Average

Calculate the company-wide average sales value.

Count how many store-category sales values are above this average.

In [ ]:
company_avg_sales = sales_data.mean()
count_above_avg = np.sum(sales_data > company_avg_sales)

print("Company-wide average sales value:", round(company_avg_sales, 2))
print("Number of store-category values above average:", count_above_avg)

Company-wide average sales value:

 389166.67
Number of store-category values above average: 10


---

## Question 85: Outlier Detection

Treat any sales value greater than:

```text
Mean + 1.5 × Standard Deviation
```

as a high-value outlier.

Identify all such sales values.

In [ ]:
outlier_threshold = sales_data.mean() + 1.5 * sales_data.std()
sales_outliers = sales_data[sales_data > outlier_threshold]

print("Outlier threshold:", round(outlier_threshold, 2))
print("High-value outlier sales figures:")
print(sales_outliers)

Outlier threshold: 734650.48
High-value outlier sales figures:
[ 850000  780000  920000 1050000]


---

# Part 15: Reshaping and Transformation

## Question 86: Flatten Sales Data

Convert the two-dimensional sales array into a one-dimensional array using:

* `flatten()`
* `ravel()`

Explain the difference between the two methods.

In [ ]:
sales_flattened = sales_data.flatten()
sales_raveled = sales_data.ravel()

print("flatten() result:", sales_flattened)
print("ravel() result:", sales_raveled)

test_ravel = sales_data.ravel()
test_ravel[0] = 999999
print("\nAfter editing ravel() output, original sales_data[0,0] is:", sales_data[0, 0])
sales_data.flat[0] = 850000
print("Restored sales_data[0,0]:", sales_data[0, 0])

print("\nDifference: flatten() always returns a copy, so editing it never affects sales_data.")
print("ravel() returns a view whenever possible, so editing it CAN change the original array,")
print("as shown above. ravel() is also faster and uses less memory since no copy is made.")

flatten() result: [ 850000  420000  310000  275000  190000  225000  780000  460000  295000
  310000  210000  240000  920000  390000  340000  260000  185000  215000
 1050000  510000  375000  345000  250000  290000  690000  405000  280000
  225000  175000  205000]
ravel() result: [ 850000  420000  310000  275000  190000  225000  780000  460000  295000
  310000  210000  240000  920000  390000  340000  260000  185000  215000
 1050000  510000  375000  345000  250000  290000  690000  405000  280000
  225000  175000  205000]

After editing ravel() output, original sales_data[0,0] is: 999999
Restored sales_data[0,0]: 850000

Difference: flatten() always returns a copy, so editing it never affects sales_data.
ravel() returns a view whenever possible, so editing it CAN change the original array,
as shown above. ravel() is also faster and uses less memory since no copy is made.


---

## Question 87: Reshape Sales Data

Reshape the `5 × 6` sales array into:

* `3 × 10`
* `10 × 3`
* `2 × 15`
* `15 × 2`

Verify that no data is lost.

In [ ]:
reshaped_3x10 = sales_data.reshape(3, 10)
reshaped_10x3 = sales_data.reshape(10, 3)
reshaped_2x15 = sales_data.reshape(2, 15)
reshaped_15x2 = sales_data.reshape(15, 2)

print("3x10:\n", reshaped_3x10)
print("\n10x3:\n", reshaped_10x3)
print("\n2x15:\n", reshaped_2x15)
print("\n15x2:\n", reshaped_15x2)

print("\nOriginal total:", sales_data.sum())
print("All reshapes preserve total sum:",
      reshaped_3x10.sum() == reshaped_10x3.sum() == reshaped_2x15.sum() == reshaped_15x2.sum() == sales_data.sum())

3x10:


 [[ 850000  420000  310000  275000  190000  225000  780000  460000  295000
   310000]
 [ 210000  240000  920000  390000  340000  260000  185000  215000 1050000
   510000]
 [ 375000  345000  250000  290000  690000  405000  280000  225000  175000
   205000]]

10x3:
 [[ 850000  420000  310000]
 [ 275000  190000  225000]
 [ 780000  460000  295000]
 [ 310000  210000  240000]
 [ 920000  390000  340000]
 [ 260000  185000  215000]
 [1050000  510000  375000]
 [ 345000  250000  290000]
 [ 690000  405000  280000]
 [ 225000  175000  205000]]

2x15:
 [[ 850000  420000  310000  275000  190000  225000  780000  460000  295000
   310000  210000  240000  920000  390000  340000]
 [ 260000  185000  215000 1050000  510000  375000  345000  250000  290000
   690000  405000  280000  225000  175000  205000]]

15x2:
 [[ 850000  420000]
 [ 310000  275000]
 [ 190000  225000]
 [ 780000  460000]
 [ 295000  310000]
 [ 210000  240000]
 [ 920000  390000]
 [ 340000  260000]
 [ 185000  215000]
 [1050000  510000]
 [ 3750

---

## Question 88: Transpose Sales Data

Transpose the sales array so that:

* Rows represent product categories
* Columns represent stores

Display the new shape.

In [ ]:
sales_transposed = sales_data.T

print("Original shape (stores x categories):", sales_data.shape)
print("Transposed shape (categories x stores):", sales_transposed.shape)
print("\nTransposed array:")
print(sales_transposed)

Original shape (stores x categories):

 (5, 6)
Transposed shape (categories x stores): (6, 5)

Transposed array:
[[ 850000  780000  920000 1050000  690000]
 [ 420000  460000  390000  510000  405000]
 [ 310000  295000  340000  375000  280000]
 [ 275000  310000  260000  345000  225000]
 [ 190000  210000  185000  250000  175000]
 [ 225000  240000  215000  290000  205000]]


---

## Question 89: Combined Sales and Cost Array

Stack `sales_data` and `cost_data` vertically.

Display the resulting shape.

In [ ]:
combined_stack = np.vstack([sales_data, cost_data])

print("Combined array shape:", combined_stack.shape)
print(combined_stack)

Combined array shape: (10, 6)
[[ 850000  420000  310000  275000  190000  225000]
 [ 780000  460000  295000  310000  210000  240000]
 [ 920000  390000  340000  260000  185000  215000]
 [1050000  510000  375000  345000  250000  290000]
 [ 690000  405000  280000  225000  175000  205000]
 [ 640000  315000  220000  205000  125000  160000]
 [ 590000  350000  210000  225000  140000  170000]
 [ 700000  295000  245000  195000  120000  155000]
 [ 790000  390000  265000  250000  165000  205000]
 [ 530000  310000  205000  170000  115000  145000]]


---

## Question 90: Store-Level Combined Data

Create a three-dimensional array where:

* First layer contains sales data
* Second layer contains cost data

Use `np.stack()`.

Display the shape of the new array.

In [ ]:
store_data_3d = np.stack([sales_data, cost_data])

print("3D array shape (layers, stores, categories):", store_data_3d.shape)
print("\nLayer 0 (Sales):")
print(store_data_3d[0])
print("\nLayer 1 (Cost):")
print(store_data_3d[1])

3D array shape (layers, stores, categories): (2, 5, 6)

Layer 0 (Sales):
[[ 850000  420000  310000  275000  190000  225000]
 [ 780000  460000  295000  310000  210000  240000]
 [ 920000  390000  340000  260000  185000  215000]
 [1050000  510000  375000  345000  250000  290000]
 [ 690000  405000  280000  225000  175000  205000]]

Layer 1 (Cost):
[[640000 315000 220000 205000 125000 160000]
 [590000 350000 210000 225000 140000 170000]
 [700000 295000 245000 195000 120000 155000]
 [790000 390000 265000 250000 165000 205000]
 [530000 310000 205000 170000 115000 145000]]


---

# Part 16: Missing Data Scenario

The company discovers that some sales values were not recorded.

```python
sales_with_missing = np.array([
    [850000, 420000, np.nan, 275000, 190000, 225000],
    [780000, 460000, 295000, np.nan, 210000, 240000],
    [920000, 390000, 340000, 260000, 185000, 215000],
    [1050000, np.nan, 375000, 345000, 250000, 290000],
    [690000, 405000, 280000, 225000, np.nan, 205000]
])
```

## Question 91: Missing Value Detection

Find:

* Total number of missing values
* Location of missing values
* Stores and categories containing missing values

In [ ]:
sales_with_missing = np.array([
    [850000, 420000, np.nan, 275000, 190000, 225000],
    [780000, 460000, 295000, np.nan, 210000, 240000],
    [920000, 390000, 340000, 260000, 185000, 215000],
    [1050000, np.nan, 375000, 345000, 250000, 290000],
    [690000, 405000, 280000, 225000, np.nan, 205000]
])

missing_mask = np.isnan(sales_with_missing)
total_missing = np.sum(missing_mask)
missing_rows, missing_cols = np.where(missing_mask)

print("Total number of missing values:", total_missing)
print("\nLocations (row, col) of missing values:")
for r, c in zip(missing_rows, missing_cols):
    print(f"({r}, {c}) -> Store: {stores[r]}, Category: {categories[c]}")

Total number of missing values: 4

Locations (row, col) of missing values:
(0, 2) -> Store: Bengaluru, Category: Clothing
(1, 3) -> Store: Hyderabad, Category: Furniture
(3, 1) -> Store: Mumbai, Category: Groceries
(4, 4) -> Store: Pune, Category: Beauty


---

## Question 92: Replace Missing Values with Zero

Replace all missing sales values with zero.

In [ ]:
sales_missing_zero = np.nan_to_num(sales_with_missing, nan=0.0)
print("Sales data with missing values replaced by zero:")
print(sales_missing_zero)

Sales data with missing values replaced by zero:
[[ 850000.  420000.       0.  275000.  190000.  225000.]
 [ 780000.  460000.  295000.       0.  210000.  240000.]
 [ 920000.  390000.  340000.  260000.  185000.  215000.]
 [1050000.       0.  375000.  345000.  250000.  290000.]
 [ 690000.  405000.  280000.  225000.       0.  205000.]]


---

## Question 93: Replace with Company Average

Replace missing values with the overall average of available sales values.

Use `np.nanmean()`.

In [ ]:
overall_avg_sales = np.nanmean(sales_with_missing)
sales_missing_overall_avg = np.where(missing_mask, overall_avg_sales, sales_with_missing)

print("Overall average of available sales values:", round(overall_avg_sales, 2))
print("\nSales data with missing values replaced by overall average:")
print(sales_missing_overall_avg)

Overall average of available sales values: 398846.15

Sales data with missing values replaced by overall average:
[[ 850000.          420000.          398846.15384615  275000.
   190000.          225000.        ]
 [ 780000.          460000.          295000.          398846.15384615
   210000.          240000.        ]
 [ 920000.          390000.          340000.          260000.
   185000.          215000.        ]
 [1050000.          398846.15384615  375000.          345000.
   250000.          290000.        ]
 [ 690000.          405000.          280000.          225000.
   398846.15384615  205000.        ]]


---

## Question 94: Replace with Category Average

Replace each missing value with the average sales value of its respective product category.

In [ ]:
category_avg_sales = np.nanmean(sales_with_missing, axis=0)

sales_missing_category_avg = sales_with_missing.copy()
row_idx, col_idx = np.where(missing_mask)
sales_missing_category_avg[row_idx, col_idx] = category_avg_sales[col_idx]

print("Category-wise average (used for replacement):")
for cat, val in zip(categories, category_avg_sales):
    print(f"{cat}: {round(val, 2)}")

print("\nSales data with missing values replaced by category average:")
print(sales_missing_category_avg)

Category-wise average (used for replacement):
Electronics: 858000.0
Groceries: 418750.0
Clothing: 322500.0
Furniture: 276250.0
Beauty: 208750.0
Sports: 235000.0

Sales data with missing values replaced by category average:
[[ 850000.  420000.  322500.  275000.  190000.  225000.]
 [ 780000.  460000.  295000.  276250.  210000.  240000.]
 [ 920000.  390000.  340000.  260000.  185000.  215000.]
 [1050000.  418750.  375000.  345000.  250000.  290000.]
 [ 690000.  405000.  280000.  225000.  208750.  205000.]]


---

## Question 95: Missing Value Impact

Compare total sales before and after replacing missing values with zero.

In [ ]:
total_before_replacement = np.nansum(sales_with_missing)  # ignores NaNs while summing
total_after_replacement = sales_missing_zero.sum()          # NaNs already replaced by 0

print("Total sales before replacement (NaNs ignored):", total_before_replacement)
print("Total sales after replacing NaNs with zero:", total_after_replacement)
print("Difference:", total_before_replacement - total_after_replacement)
print("\nNote: replacing NaN with 0 and then summing gives the same result as summing")
print("while ignoring NaNs (nansum) - both approaches simply exclude the missing entries.")
print("A plain .sum() on data containing NaN, however, would return NaN:", sales_with_missing.sum())

Total sales before replacement (NaNs ignored): 10370000.0
Total sales after replacing NaNs with zero: 10370000.0
Difference: 0.0

Note: replacing NaN with 0 and then summing gives the same result as summing
while ignoring NaNs (nansum) - both approaches simply exclude the missing entries.
A plain .sum() on data containing NaN, however, would return NaN: nan


---

# Part 17: Conditional Business Decisions

## Question 96: Store Performance Labels

Classify each store based on total sales:

* `"Excellent"`: Total sales above ₹28,00,000
* `"Good"`: Total sales between ₹24,00,000 and ₹28,00,000
* `"Needs Improvement"`: Total sales below ₹24,00,000

Use `np.select()`.

In [ ]:
store_performance_labels = np.select(
    [store_wise_sales > 2800000, store_wise_sales >= 2400000],
    ["Excellent", "Good"],
    default="Needs Improvement"
)

for st, sales_val, label in zip(stores, store_wise_sales, store_performance_labels):
    print(f"{st}: Total Sales = {sales_val} -> {label}")

Bengaluru: Total Sales = 2270000 -> Needs Improvement
Hyderabad: Total Sales = 2295000 -> Needs Improvement
Chennai: Total Sales = 2310000 -> Needs Improvement
Mumbai: Total Sales = 2820000 -> Excellent
Pune: Total Sales = 1980000 -> Needs Improvement


---

## Question 97: Profitability Classification

Classify each store-category combination as:

* `"Highly Profitable"`: Profit above ₹2,00,000
* `"Profitable"`: Profit between ₹1,00,000 and ₹2,00,000
* `"Low Profit"`: Profit below ₹1,00,000
* `"Loss"`: Profit below zero

In [ ]:
profitability_classification = np.select(
    [profit_data < 0, profit_data < 100000, profit_data <= 200000],
    ["Loss", "Low Profit", "Profitable"],
    default="Highly Profitable"
)

print("Profitability classification matrix:")
print(profitability_classification)

Profitability classification matrix:
[['Highly Profitable' 'Profitable' 'Low Profit' 'Low Profit' 'Low Profit'
  'Low Profit']
 ['Profitable' 'Profitable' 'Low Profit' 'Low Profit' 'Low Profit'
  'Low Profit']
 ['Highly Profitable' 'Low Profit' 'Low Profit' 'Low Profit' 'Low Profit'
  'Low Profit']
 ['Highly Profitable' 'Profitable' 'Profitable' 'Low Profit' 'Low Profit'
  'Low Profit']
 ['Profitable' 'Low Profit' 'Low Profit' 'Low Profit' 'Low Profit'
  'Low Profit']]


---

## Question 98: Expansion Recommendation

A store is eligible for expansion when:

* Total sales are above ₹25,00,000
* Total profit is above ₹7,00,000
* Overall profit margin is above 25%

Identify eligible stores.

In [ ]:
expansion_mask = (store_wise_sales > 2500000) & (store_wise_profit > 700000) & (store_profit_margin > 25)
expansion_eligible_stores = stores[expansion_mask]

print("Stores eligible for expansion:")
print(expansion_eligible_stores if expansion_eligible_stores.size else "None")

Stores eligible for expansion:
['Mumbai']


---

## Question 99: Category Promotion Recommendation

A product category should receive additional promotion when:

* Average sales are below ₹3,00,000
* Profit margin is above 25%

Identify the categories that satisfy these conditions.

In [ ]:
promotion_mask = (avg_sales_per_category < 300000) & (category_profit_margin > 25)
promotion_categories = categories[promotion_mask]

print("Categories recommended for additional promotion:")
print(promotion_categories if promotion_categories.size else "None")

Categories recommended for additional promotion:
['Furniture' 'Beauty' 'Sports']


---

## Question 100: Cost Optimization Recommendation

Find store-category combinations where:

* Cost is greater than ₹2,50,000
* Profit margin is below 25%

These combinations should be reviewed for cost optimization.

In [ ]:
cost_optimization_mask = (cost_data > 250000) & (profit_margin < 25)
cost_opt_store_idx, cost_opt_cat_idx = np.where(cost_optimization_mask)

print("Store-category combinations needing cost optimization:")
for r, c in zip(cost_opt_store_idx, cost_opt_cat_idx):
    print(f"Store: {stores[r]}, Category: {categories[c]}, "
          f"Cost: {cost_data[r, c]}, Margin: {profit_margin[r, c]}%")

Store-category combinations needing cost optimization:
Store: Bengaluru, Category: Electronics, Cost: 640000, Margin: 24.71%
Store: Hyderabad, Category: Electronics, Cost: 590000, Margin: 24.36%
Store: Hyderabad, Category: Groceries, Cost: 350000, Margin: 23.91%
Store: Chennai, Category: Electronics, Cost: 700000, Margin: 23.91%
Store: Chennai, Category: Groceries, Cost: 295000, Margin: 24.36%
Store: Mumbai, Category: Electronics, Cost: 790000, Margin: 24.76%
Store: Mumbai, Category: Groceries, Cost: 390000, Margin: 23.53%
Store: Pune, Category: Electronics, Cost: 530000, Margin: 23.19%
Store: Pune, Category: Groceries, Cost: 310000, Margin: 23.46%


---

# Part 18: Inventory Scenario

Use the following inventory arrays:

```python
opening_stock = np.array([
    [120, 450, 300, 90, 250, 180],
    [110, 500, 280, 100, 270, 200],
    [130, 420, 320, 85, 240, 170],
    [150, 550, 350, 120, 300, 220],
    [100, 400, 260, 75, 220, 160]
])

units_sold = np.array([
    [95, 380, 240, 65, 190, 140],
    [85, 420, 230, 75, 210, 155],
    [105, 350, 270, 60, 180, 130],
    [125, 470, 295, 90, 235, 175],
    [75, 330, 205, 55, 165, 120]
])
```

## Question 101: Closing Stock

Calculate closing stock using:

```text
Closing Stock = Opening Stock - Units Sold
```

In [ ]:
opening_stock = np.array([
    [120, 450, 300, 90, 250, 180],
    [110, 500, 280, 100, 270, 200],
    [130, 420, 320, 85, 240, 170],
    [150, 550, 350, 120, 300, 220],
    [100, 400, 260, 75, 220, 160]
])

units_sold = np.array([
    [95, 380, 240, 65, 190, 140],
    [85, 420, 230, 75, 210, 155],
    [105, 350, 270, 60, 180, 130],
    [125, 470, 295, 90, 235, 175],
    [75, 330, 205, 55, 165, 120]
])

closing_stock = opening_stock - units_sold
print("Closing stock matrix:")
print(closing_stock)

Closing stock matrix:


[[25 70 60 25 60 40]
 [25 80 50 25 60 45]
 [25 70 50 25 60 40]
 [25 80 55 30 65 45]
 [25 70 55 20 55 40]]


---

## Question 102: Low-Stock Identification

Identify store-category combinations where closing stock is below 30 units.

In [ ]:
low_stock_mask = closing_stock < 30
low_stock_store_idx, low_stock_cat_idx = np.where(low_stock_mask)

print("Low-stock (< 30 units) combinations:")
for r, c in zip(low_stock_store_idx, low_stock_cat_idx):
    print(f"Store: {stores[r]}, Category: {categories[c]}, Closing Stock: {closing_stock[r, c]}")

Low-stock (< 30 units) combinations:
Store: Bengaluru, Category: Electronics, Closing Stock: 25
Store: Bengaluru, Category: Furniture, Closing Stock: 25
Store: Hyderabad, Category: Electronics, Closing Stock: 25
Store: Hyderabad, Category: Furniture, Closing Stock: 25
Store: Chennai, Category: Electronics, Closing Stock: 25
Store: Chennai, Category: Furniture, Closing Stock: 25
Store: Mumbai, Category: Electronics, Closing Stock: 25
Store: Pune, Category: Electronics, Closing Stock: 25
Store: Pune, Category: Furniture, Closing Stock: 20


---

## Question 103: Stock Availability Percentage

Calculate the percentage of opening stock that remains unsold.

In [ ]:
stock_availability_pct = np.round((closing_stock / opening_stock) * 100, 2)
print("Percentage of opening stock remaining unsold:")
print(stock_availability_pct)

Percentage of opening stock remaining unsold:
[[20.83 15.56 20.   27.78 24.   22.22]
 [22.73 16.   17.86 25.   22.22 22.5 ]
 [19.23 16.67 15.62 29.41 25.   23.53]
 [16.67 14.55 15.71 25.   21.67 20.45]
 [25.   17.5  21.15 26.67 25.   25.  ]]


---

## Question 104: Fast-Moving Products

A product is considered fast-moving when more than 80% of its opening stock has been sold.

Identify fast-moving store-category combinations.

In [ ]:
sell_through_pct = units_sold / opening_stock
fast_moving_mask = sell_through_pct > 0.80
fast_moving_store_idx, fast_moving_cat_idx = np.where(fast_moving_mask)

print("Fast-moving store-category combinations (>80% sold):")
for r, c in zip(fast_moving_store_idx, fast_moving_cat_idx):
    print(f"Store: {stores[r]}, Category: {categories[c]}, "
          f"Sold: {round(sell_through_pct[r, c] * 100, 2)}%")

Fast-moving store-category combinations (>80% sold):
Store: Bengaluru, Category: Groceries, Sold: 84.44%
Store: Hyderabad, Category: Groceries, Sold: 84.0%
Store: Hyderabad, Category: Clothing, Sold: 82.14%
Store: Chennai, Category: Electronics, Sold: 80.77%
Store: Chennai, Category: Groceries, Sold: 83.33%
Store: Chennai, Category: Clothing, Sold: 84.38%
Store: Mumbai, Category: Electronics, Sold: 83.33%
Store: Mumbai, Category: Groceries, Sold: 85.45%
Store: Mumbai, Category: Clothing, Sold: 84.29%
Store: Pune, Category: Groceries, Sold: 82.5%


---

## Question 105: Category-Wise Units Sold

Calculate the total units sold for each product category.

In [ ]:
category_units_sold = units_sold.sum(axis=0)

for cat, val in zip(categories, category_units_sold):
    print(f"{cat}: {val} units")

Electronics: 485 units
Groceries: 1950 units
Clothing: 1240 units
Furniture: 345 units
Beauty: 980 units
Sports: 720 units


---

## Question 106: Store-Wise Units Sold

Calculate total units sold by each store.

In [ ]:
store_units_sold = units_sold.sum(axis=1)

for st, val in zip(stores, store_units_sold):
    print(f"{st}: {val} units")

Bengaluru: 1110 units
Hyderabad: 1175 units
Chennai: 1095 units
Mumbai: 1390 units
Pune: 950 units


---

## Question 107: Best Store by Unit Sales

Identify the store that sold the highest total number of units.

In [ ]:
best_unit_store_idx = np.argmax(store_units_sold)
print("Store with highest total units sold:", stores[best_unit_store_idx])
print("Units sold:", store_units_sold[best_unit_store_idx])

Store with highest total units sold: Mumbai
Units sold: 1390


---

## Question 108: Reorder Requirement

A store should reorder a category when closing stock is below 25 units.

Create a Boolean reorder matrix.

In [ ]:
reorder_mask = closing_stock < 25
print("Reorder requirement matrix (True = needs reorder):")
print(reorder_mask)

Reorder requirement matrix (True = needs reorder):
[[False False False False False False]
 [False False False False False False]
 [False False False False False False]
 [False False False False False False]
 [False False False  True False False]]


---

## Question 109: Reorder Quantity

Assume each category must maintain 100 units after reordering.

Calculate reorder quantities only for combinations where closing stock is below 25.

For all other combinations, return zero.

In [ ]:
reorder_quantity = np.where(reorder_mask, 100 - closing_stock, 0)
print("Reorder quantity matrix:")
print(reorder_quantity)

Reorder quantity matrix:
[[ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0 80  0  0]]


---

# Part 19: Advanced NumPy Questions

## Question 110: Cumulative Store Sales

Calculate cumulative category sales for each store using `np.cumsum()`.

In [ ]:
cumulative_store_sales = np.cumsum(sales_data, axis=1)
print("Cumulative sales across categories, per store:")
print(cumulative_store_sales)

Cumulative sales across categories, per store:
[[ 850000 1270000 1580000 1855000 2045000 2270000]
 [ 780000 1240000 1535000 1845000 2055000 2295000]
 [ 920000 1310000 1650000 1910000 2095000 2310000]
 [1050000 1560000 1935000 2280000 2530000 2820000]
 [ 690000 1095000 1375000 1600000 1775000 1980000]]


---

## Question 111: Cumulative Company Sales

Flatten the sales array and calculate cumulative company sales.

In [ ]:
cumulative_company_sales = np.cumsum(sales_data.flatten())
print("Cumulative company-wide sales:")
print(cumulative_company_sales)

Cumulative company-wide sales:
[  850000  1270000  1580000  1855000  2045000  2270000  3050000  3510000
  3805000  4115000  4325000  4565000  5485000  5875000  6215000  6475000
  6660000  6875000  7925000  8435000  8810000  9155000  9405000  9695000
 10385000 10790000 11070000 11295000 11470000 11675000]


---

## Question 112: Sales Difference Between Categories

Calculate the difference between consecutive category sales for each store using `np.diff()`.

In [ ]:
category_diff = np.diff(sales_data, axis=1)
print("Difference between consecutive category sales, per store:")
print(category_diff)

Difference between consecutive category sales, per store:
[[-430000 -110000  -35000  -85000   35000]
 [-320000 -165000   15000 -100000   30000]
 [-530000  -50000  -80000  -75000   30000]
 [-540000 -135000  -30000  -95000   40000]
 [-285000 -125000  -55000  -50000   30000]]


---

## Question 113: Unique Sales Values

Find all unique sales values in `sales_data`.

Also count the frequency of each value.

In [ ]:
unique_sales_values, unique_counts = np.unique(sales_data, return_counts=True)

print("Unique sales values and their frequency:")
for val, cnt in zip(unique_sales_values, unique_counts):
    print(f"{val}: {cnt} time(s)")

Unique sales values and their frequency:
175000: 1 time(s)
185000: 1 time(s)
190000: 1 time(s)
205000: 1 time(s)
210000: 1 time(s)
215000: 1 time(s)
225000: 2 time(s)
240000: 1 time(s)
250000: 1 time(s)
260000: 1 time(s)
275000: 1 time(s)
280000: 1 time(s)
290000: 1 time(s)
295000: 1 time(s)
310000: 2 time(s)
340000: 1 time(s)
345000: 1 time(s)
375000: 1 time(s)
390000: 1 time(s)
405000: 1 time(s)
420000: 1 time(s)
460000: 1 time(s)
510000: 1 time(s)
690000: 1 time(s)
780000: 1 time(s)
850000: 1 time(s)
920000: 1 time(s)
1050000: 1 time(s)


---

## Question 114: Search for Specific Sales

Check whether the value ₹4,20,000 exists in the sales array.

Return its location.

In [ ]:
search_value = 420000
value_exists = np.isin(search_value, sales_data)
search_rows, search_cols = np.where(sales_data == search_value)

print(f"Does {search_value} exist in sales_data?", value_exists)
print("Location(s) (row, col):", list(zip(search_rows.tolist(), search_cols.tolist())))
for r, c in zip(search_rows, search_cols):
    print(f"Store: {stores[r]}, Category: {categories[c]}")

Does 420000 exist in sales_data? True
Location(s) (row, col): [(0, 1)]
Store: Bengaluru, Category: Groceries


---

## Question 115: Compare Sales and Cost

Create a Boolean array showing whether sales are greater than cost for every combination.

Verify whether all combinations are profitable using `np.all()`.

In [ ]:
sales_gt_cost = sales_data > cost_data
all_profitable = np.all(sales_gt_cost)

print("Sales greater than cost matrix:")
print(sales_gt_cost)
print("\nAre ALL store-category combinations profitable?", all_profitable)

Sales greater than cost matrix:


[[ True  True  True  True  True  True]
 [ True  True  True  True  True  True]
 [ True  True  True  True  True  True]
 [ True  True  True  True  True  True]
 [ True  True  True  True  True  True]]

Are ALL store-category combinations profitable? True


---

## Question 116: Clip Profit Values

For reporting purposes, management wants profit values limited to:

* Minimum: ₹75,000
* Maximum: ₹2,50,000

Use `np.clip()` to create a reporting profit array.

In [ ]:
reporting_profit = np.clip(profit_data, 75000, 250000)
print("Original profit matrix:")
print(profit_data)
print("\nClipped reporting profit matrix (75,000 to 250,000):")
print(reporting_profit)

Original profit matrix:


[[210000 105000  90000  70000  65000  65000]
 [190000 110000  85000  85000  70000  70000]
 [220000  95000  95000  65000  65000  60000]
 [260000 120000 110000  95000  85000  85000]
 [160000  95000  75000  55000  60000  60000]]

Clipped reporting profit matrix (75,000 to 250,000):
[[210000 105000  90000  75000  75000  75000]
 [190000 110000  85000  85000  75000  75000]
 [220000  95000  95000  75000  75000  75000]
 [250000 120000 110000  95000  85000  85000]
 [160000  95000  75000  75000  75000  75000]]


---

## Question 117: Normalize Sales Data

Normalize all sales values using min-max normalization:

```text
Normalized Sales =
(Sales - Minimum Sales) / (Maximum Sales - Minimum Sales)
```

In [ ]:
normalized_sales = (sales_data - sales_data.min()) / (sales_data.max() - sales_data.min())
print("Min-max normalized sales data:")
print(np.round(normalized_sales, 4))

Min-max normalized sales data:
[[0.7714 0.28   0.1543 0.1143 0.0171 0.0571]
 [0.6914 0.3257 0.1371 0.1543 0.04   0.0743]
 [0.8514 0.2457 0.1886 0.0971 0.0114 0.0457]
 [1.     0.3829 0.2286 0.1943 0.0857 0.1314]
 [0.5886 0.2629 0.12   0.0571 0.     0.0343]]


---

## Question 118: Standardize Sales Data

Standardize sales using:

```text
Standardized Sales =
(Sales - Mean Sales) / Standard Deviation
```

In [ ]:
standardized_sales = (sales_data - sales_data.mean()) / sales_data.std()
print("Standardized (z-score) sales data:")
print(np.round(standardized_sales, 4))

Standardized (z-score) sales data:
[[ 2.0008  0.1339 -0.3437 -0.4957 -0.8647 -0.7128]
 [ 1.6969  0.3075 -0.4088 -0.3437 -0.7779 -0.6476]
 [ 2.3047  0.0036 -0.2135 -0.5608 -0.8864 -0.7562]
 [ 2.8692  0.5246 -0.0615 -0.1918 -0.6042 -0.4306]
 [ 1.3061  0.0687 -0.474  -0.7128 -0.9299 -0.7996]]


---

## Question 119: Round Values

Convert profit margin values into:

* Two decimal places
* Nearest whole number
* Floor values
* Ceiling values

In [ ]:
margin_2dp = np.round(profit_margin, 2)
margin_nearest_whole = np.rint(profit_margin)
margin_floor = np.floor(profit_margin)
margin_ceil = np.ceil(profit_margin)

print("Profit margin - 2 decimal places:\n", margin_2dp)
print("\nProfit margin - nearest whole number:\n", margin_nearest_whole)
print("\nProfit margin - floor values:\n", margin_floor)
print("\nProfit margin - ceiling values:\n", margin_ceil)

Profit margin - 2 decimal places:
 [[24.71 25.   29.03 25.45 34.21 28.89]
 [24.36 23.91 28.81 27.42 33.33 29.17]
 [23.91 24.36 27.94 25.   35.14 27.91]
 [24.76 23.53 29.33 27.54 34.   29.31]
 [23.19 23.46 26.79 24.44 34.29 29.27]]

Profit margin - nearest whole number:
 [[25. 25. 29. 25. 34. 29.]
 [24. 24. 29. 27. 33. 29.]
 [24. 24. 28. 25. 35. 28.]
 [25. 24. 29. 28. 34. 29.]
 [23. 23. 27. 24. 34. 29.]]

Profit margin - floor values:
 [[24. 25. 29. 25. 34. 28.]
 [24. 23. 28. 27. 33. 29.]
 [23. 24. 27. 25. 35. 27.]
 [24. 23. 29. 27. 34. 29.]
 [23. 23. 26. 24. 34. 29.]]

Profit margin - ceiling values:
 [[25. 25. 30. 26. 35. 29.]
 [25. 24. 29. 28. 34. 30.]
 [24. 25. 28. 25. 36. 28.]
 [25. 24. 30. 28. 34. 30.]
 [24. 24. 27. 25. 35. 30.]]


---

## Question 120: Save and Load Data

Save the following arrays as NumPy binary files:

* `sales_data.npy`
* `cost_data.npy`
* `profit_data.npy`

Load them back and verify that the loaded arrays are equal to the original arrays.

In [ ]:
np.save("sales_data.npy", sales_data)
np.save("cost_data.npy", cost_data)
np.save("profit_data.npy", profit_data)

loaded_sales = np.load("sales_data.npy")
loaded_cost = np.load("cost_data.npy")
loaded_profit = np.load("profit_data.npy")

print("Sales data matches after save/load:", np.array_equal(sales_data, loaded_sales))
print("Cost data matches after save/load:", np.array_equal(cost_data, loaded_cost))
print("Profit data matches after save/load:", np.array_equal(profit_data, loaded_profit))

Sales data matches after save/load: True
Cost data matches after save/load: True
Profit data matches after save/load: True


---

# Part 20: Final Retail Analytics Project

Prepare a complete NumPy-based retail performance report.

## Company Summary

In [ ]:
# ----- Company Summary -----
company_total_sales = sales_data.sum()
company_total_cost = cost_data.sum()
company_total_profit = profit_data.sum()
company_profit_margin = round((company_total_profit / company_total_sales) * 100, 2)
avg_transaction_value = round(sales_data.mean(), 2)
highest_transaction = sales_data.max()
lowest_transaction = sales_data.min()

print("Total Company Sales:", company_total_sales)
print("Total Company Cost:", company_total_cost)
print("Total Company Profit:", company_total_profit)
print("Overall Company Profit Margin:", company_profit_margin, "%")
print("Average Transaction Value:", avg_transaction_value)
print("Highest Individual Sales Value:", highest_transaction)
print("Lowest Individual Sales Value:", lowest_transaction)

Total Company Sales:

 11675000
Total Company Cost: 8600000
Total Company Profit: 3075000
Overall Company Profit Margin: 26.34 %
Average Transaction Value: 389166.67
Highest Individual Sales Value: 1050000
Lowest Individual Sales Value: 175000


## Store Summary

In [ ]:
# ----- Store Summary -----
# Expansion eligibility reuses the Question 98 rule
store_expansion_eligible = expansion_mask

print(f"{'Store':<12}{'Sales':>10}{'Cost':>10}{'Profit':>10}{'Margin%':>10}{'Targets':>10}{'Units':>10}{'Expand':>10}")
for i, st in enumerate(stores):
    print(f"{st:<12}{store_wise_sales[i]:>10}{store_wise_cost[i]:>10}{store_wise_profit[i]:>10}"
          f"{store_profit_margin[i]:>10}{achievement_count[i]:>10}{store_units_sold[i]:>10}"
          f"{str(bool(store_expansion_eligible[i])):>10}")

Store            Sales      Cost    Profit   Margin%   Targets     Units    Expand
Bengaluru      2270000   1665000    605000     26.65         1      1110     False
Hyderabad      2295000   1685000    610000     26.58         4      1175     False
Chennai        2310000   1710000    600000     25.97         2      1095     False
Mumbai         2820000   2065000    755000     26.77         6      1390      True
Pune           1980000   1475000    505000     25.51         0       950     False


## Category Summary

In [ ]:
# ----- Category Summary -----
print(f"{'Category':<15}{'Sales':>10}{'Cost':>10}{'Profit':>10}{'Margin%':>10}{'AvgSales':>12}{'Stores@Tgt':>12}{'Units':>10}")
for i, cat in enumerate(categories):
    print(f"{cat:<15}{category_wise_sales[i]:>10}{category_wise_cost[i]:>10}{category_wise_profit[i]:>10}"
          f"{category_profit_margin[i]:>10}{avg_sales_per_category[i]:>12.2f}"
          f"{achievement_by_category[i]:>12}{category_units_sold[i]:>10}")

Category            Sales      Cost    Profit   Margin%    AvgSales  Stores@Tgt     Units


Electronics       4290000   3250000   1040000     24.24   858000.00           3       485
Groceries         2185000   1660000    525000     24.03   437000.00           2      1950
Clothing          1600000   1145000    455000     28.44   320000.00           2      1240
Furniture         1415000   1045000    370000     26.15   283000.00           2       345
Beauty            1010000    665000    345000     34.16   202000.00           2       980
Sports            1175000    835000    340000     28.94   235000.00           2       720


## Management Recommendations

In [ ]:
# ----- Management Recommendations (derived from the analysis above) -----
best_performing_store = stores[most_profitable_store_idx]
lowest_performing_store = stores[least_profitable_store_idx]
most_profitable_cat = categories[most_profitable_category_idx]
lowest_profit_cat = categories[least_profitable_category_idx]
expansion_recommendation = expansion_eligible_stores
marketing_recommendation = promotion_categories
cost_review_combinations = list(zip(stores[cost_opt_store_idx].tolist(), categories[cost_opt_cat_idx].tolist()))
reorder_combinations = list(zip(
    stores[np.where(reorder_mask)[0]].tolist(),
    categories[np.where(reorder_mask)[1]].tolist()
))

print("Best-performing store (by profit):", best_performing_store)
print("Lowest-performing store (by profit):", lowest_performing_store)
print("Most profitable category:", most_profitable_cat)
print("Lowest-profit category:", lowest_profit_cat)
print("Store(s) recommended for expansion:", expansion_recommendation.tolist() if len(expansion_recommendation) else "None")
print("Category/categories needing marketing:", marketing_recommendation.tolist() if len(marketing_recommendation) else "None")
print("Store-category combos needing cost optimization:", cost_review_combinations)
print("Store-category combos needing inventory reorder:", reorder_combinations)

Best-performing store (by profit): Mumbai
Lowest-performing store (by profit): Pune
Most profitable category: Electronics
Lowest-profit category: Sports
Store(s) recommended for expansion: ['Mumbai']
Category/categories needing marketing: ['Furniture', 'Beauty', 'Sports']
Store-category combos needing cost optimization: [('Bengaluru', 'Electronics'), ('Hyderabad', 'Electronics'), ('Hyderabad', 'Groceries'), ('Chennai', 'Electronics'), ('Chennai', 'Groceries'), ('Mumbai', 'Electronics'), ('Mumbai', 'Groceries'), ('Pune', 'Electronics'), ('Pune', 'Groceries')]
Store-category combos needing inventory reorder: [('Pune', 'Furniture')]


## Final Formatted Report

Matches the "Expected Final Output Format" template from the assignment.

In [ ]:
print("SMARTMART RETAIL PERFORMANCE REPORT")
print("===================================")

print("\nCOMPANY SUMMARY")
print("---------------")
print(f"Total Sales: {company_total_sales}")
print(f"Total Cost: {company_total_cost}")
print(f"Total Profit: {company_total_profit}")
print(f"Overall Profit Margin: {company_profit_margin}%")

print("\nBEST-PERFORMING STORE")
print("---------------------")
print(f"Store: {stores[most_profitable_store_idx]}")
print(f"Sales: {store_wise_sales[most_profitable_store_idx]}")
print(f"Cost: {store_wise_cost[most_profitable_store_idx]}")
print(f"Profit: {store_wise_profit[most_profitable_store_idx]}")
print(f"Profit Margin: {store_profit_margin[most_profitable_store_idx]}%")

print("\nMOST PROFITABLE CATEGORY")
print("------------------------")
print(f"Category: {categories[most_profitable_category_idx]}")
print(f"Sales: {category_wise_sales[most_profitable_category_idx]}")
print(f"Cost: {category_wise_cost[most_profitable_category_idx]}")
print(f"Profit: {category_wise_profit[most_profitable_category_idx]}")
print(f"Profit Margin: {category_profit_margin[most_profitable_category_idx]}%")

print("\nEXPANSION RECOMMENDATIONS")
print("-------------------------")
print(f"Eligible Stores: {expansion_eligible_stores.tolist() if len(expansion_eligible_stores) else 'None'}")

print("\nCOST OPTIMIZATION")
print("-----------------")
if cost_opt_store_idx.size:
    for r, c in zip(cost_opt_store_idx, cost_opt_cat_idx):
        print(f"Store: {stores[r]}")
        print(f"Category: {categories[c]}")
        print(f"Cost: {cost_data[r, c]}")
        print(f"Profit Margin: {profit_margin[r, c]}%")
        print("-")
else:
    print("None")

print("\nINVENTORY REORDER")
print("-----------------")
reorder_rows, reorder_cols = np.where(reorder_mask)
if reorder_rows.size:
    for r, c in zip(reorder_rows, reorder_cols):
        print(f"Store: {stores[r]}")
        print(f"Category: {categories[c]}")
        print(f"Closing Stock: {closing_stock[r, c]}")
        print(f"Reorder Quantity: {reorder_quantity[r, c]}")
        print("-")
else:
    print("None")

SMARTMART RETAIL PERFORMANCE REPORT



COMPANY SUMMARY
---------------
Total Sales: 11675000
Total Cost: 8600000
Total Profit: 3075000
Overall Profit Margin: 26.34%

BEST-PERFORMING STORE
---------------------
Store: Mumbai
Sales: 2820000
Cost: 2065000
Profit: 755000
Profit Margin: 26.77%

MOST PROFITABLE CATEGORY
------------------------
Category: Electronics
Sales: 4290000
Cost: 3250000
Profit: 1040000
Profit Margin: 24.24%

EXPANSION RECOMMENDATIONS
-------------------------
Eligible Stores: ['Mumbai']

COST OPTIMIZATION
-----------------
Store: Bengaluru
Category: Electronics
Cost: 640000
Profit Margin: 24.71%
-
Store: Hyderabad
Category: Electronics
Cost: 590000
Profit Margin: 24.36%
-
Store: Hyderabad
Category: Groceries
Cost: 350000
Profit Margin: 23.91%
-
Store: Chennai
Category: Electronics
Cost: 700000
Profit Margin: 23.91%
-
Store: Chennai
Category: Groceries
Cost: 295000
Profit Margin: 24.36%
-
Store: Mumbai
Category: Electronics
Cost: 790000
Profit Margin: 24.76%
-
Store: Mumbai
Category: Groceries
Cost: 3900

---

# Business Observations

Based on the complete NumPy analysis above, five key observations emerge:

1. **Mumbai is the standout performer.** It has the highest total sales (₹28,20,000), the highest total profit (₹7,55,000), the best profit margin (26.77%), and is the only store that hit its sales target in all six categories — making it the single store that qualifies for expansion under management's criteria (sales > ₹25,00,000, profit > ₹7,00,000, margin > 25%).

2. **Pune is the clear laggard.** Its total sales (₹19,80,000) and total profit (₹5,05,000) are the lowest of all five stores, and it is the only store that achieved **zero** of its six category targets, despite still posting a positive margin (25.51%) — the issue is under-selling against target, not unprofitability.

3. **Revenue and margin don't move together.** Electronics drives the most revenue by far (₹42,90,000, ~37% of company sales) but has one of the weaker margins (24.24%), while Beauty is the smallest category by revenue (₹10,10,000) yet the most profitable per rupee sold (34.16% margin). SmartMart is volume-heavy in a comparatively low-margin category.

4. **Electronics and Groceries are structural cost problems, not store-specific ones.** These two categories account for 9 of the store-category combinations flagged for cost optimization (cost > ₹2,50,000 and margin < 25%) — nearly every store has both categories flagged, pointing to a company-wide sourcing/logistics issue rather than a single underperforming location.

5. **Store profit margins are tightly clustered (25.5%–26.8%) even though target achievement varies enormously (0 to 6 out of 6 targets met).** This means the target-attainment gap is being driven by category-specific demand or target-setting issues rather than by store-level pricing or cost discipline — every store is running the business efficiently on the sales it does make.

# Management Recommendations

Based on the analysis, three (plus one bonus) actionable recommendations:

1. **Prioritize Mumbai for the next expansion.** It is the only store meeting all three expansion criteria (sales, profit, and margin). Study its category mix and local execution and use it as the template for the next store opening.

2. **Launch a targeted turnaround plan for Pune.** It missed every single category target and trails on both sales and profit. Since its margin is comparable to the other stores, the fix should focus on demand generation (local marketing, footfall, assortment) rather than cost-cutting.

3. **Renegotiate sourcing for Electronics and Groceries company-wide.** Because these two categories are flagged for cost optimization at nearly every store, a centralized vendor/logistics renegotiation is likely to be far more effective than five separate store-level fixes.

4. **Shift promotional spend toward Furniture, Beauty, and Sports.** These categories already clear a 25%+ margin but have below-average sales volume, so incremental marketing rupees spent here should convert to profit more efficiently than in the already-high-volume Electronics category.